consensus/spos/worker.go

```diff
@@ -855,66 +848,25 @@ func (wrk *Worker) verifyEquivalentMessageSignature(cnsMsg *consensus.Message) e
 	return wrk.headerSigVerifier.VerifySignatureForHash(wrk.consensusState.Header, cnsMsg.BlockHeaderHash, cnsMsg.PubKeysBitmap, cnsMsg.AggregateSignature)
 }
 
-func (wrk *Worker) processInvalidEquivalentMessageUnprotected(blockHeaderHash []byte) {
-	hdrHash := string(blockHeaderHash)
-	delete(wrk.equivalentMessages, hdrHash)
-}
-
-// getEquivalentMessages returns a copy of the equivalent messages
-func (wrk *Worker) getEquivalentMessages() map[string]*consensus.EquivalentMessageInfo {
-	wrk.mutEquivalentMessages.RLock()
-	defer wrk.mutEquivalentMessages.RUnlock()
-
-	equivalentMessagesCopy := make(map[string]*consensus.EquivalentMessageInfo, len(wrk.equivalentMessages))
-	for hash, cnt := range wrk.equivalentMessages {
-		equivalentMessagesCopy[hash] = cnt
-	}
-
-	return equivalentMessagesCopy
-}
-
 // HasEquivalentMessage returns true if an equivalent message was received before
 func (wrk *Worker) HasEquivalentMessage(headerHash []byte) bool {
-	wrk.mutEquivalentMessages.RLock()
-	defer wrk.mutEquivalentMessages.RUnlock()
-
-	info, has := wrk.equivalentMessages[string(headerHash)]
-
-	return has && info.Validated
+	_, err := wrk.GetEquivalentProof(headerHash)
+	return err == nil
 }
 
 // GetEquivalentProof returns the equivalent proof for the provided hash
-func (wrk *Worker) GetEquivalentProof(headerHash []byte) (data.HeaderProof, error) {
-	wrk.mutEquivalentMessages.RLock()
-	defer wrk.mutEquivalentMessages.RUnlock()
-
-	info, has := wrk.equivalentMessages[string(headerHash)]
-	if !has {
-		return data.HeaderProof{}, ErrMissingEquivalentProof
-	}
-
-	if !info.Validated {
-		return data.HeaderProof{}, ErrEquivalentProofNotValidated
-	}
-
-	return info.Proof, nil
+func (wrk *Worker) GetEquivalentProof(headerHash []byte) (data.HeaderProofHandler, error) {
+	return wrk.equivalentProofsPool.GetNotarizedProof(wrk.shardCoordinator.SelfId(), headerHash)
 }
 
-// SetValidEquivalentProof saves the equivalent proof for the provided header and marks it as validated
-func (wrk *Worker) SetValidEquivalentProof(headerHash []byte, proof data.HeaderProof) {
-	wrk.mutEquivalentMessages.Lock()
-	defer wrk.mutEquivalentMessages.Unlock()
-
-	hash := string(headerHash)
-	equivalentMessage, ok := wrk.equivalentMessages[hash]
-	if !ok {
-		equivalentMessage = &consensus.EquivalentMessageInfo{
-			NumMessages: 1,
-		}
-		wrk.equivalentMessages[hash] = equivalentMessage
+// SetValidEquivalentProof saves the equivalent proof for the provided header
+func (wrk *Worker) SetValidEquivalentProof(proof data.HeaderProofHandler) {
```

Comment:
we should unlink the worker with the equivalent proofs and only pass the equivalentProofsPool to subroundBlock and subroundEndround

Reply:
done